python-checks/src/main/java/org/sonar/python/checks/SpecialMethodReturnTypeCheck.java

```diff
@@ -0,0 +1,361 @@
+      ctx.addIssue(yieldKeyword, String.format(GENERATOR_METHOD_MESSAGE, expectedReturnType));
+    }
+
+    final List<ReturnStatement> returnStmts = returnStmtCollector.getReturnStmts();
+    if (returnStmts.isEmpty() && yieldKeywords.isEmpty() && !returnStmtCollector.raisesWhitelistingException()) {
+      ctx.addIssue(funDef.defKeyword(), funDef.colon(), String.format(NO_RETURN_STMTS_MESSAGE, expectedReturnType));
+      return;
+    }
+
+    for (final ReturnStatement returnStmt : returnStmts) {
+      checkReturnStmt(ctx, funNameString, expectedReturnType, returnStmt);
+    }
+  }
+
+  private static void checkForAsync(SubscriptionContext ctx, FunctionDef funDef, String expectedReturnType) {
+    final Token asyncKeyword = funDef.asyncKeyword();
+    if (asyncKeyword != null) {
+      ctx.addIssue(asyncKeyword, String.format(COROUTINE_METHOD_MESSAGE, expectedReturnType));
+    }
+  }
+
+  /**
+   * {@code checkReturnStmt} inspects the expressions contained in a return statement against the given {@code expectedReturnType}.
+   * Some additional checks are performed if {@code methodName} is {@code "__getnewargs_ex__"}.
+   *
+   * To avoid triggering too many false positives, we perform rather weak type checks using {@code canBeOrExtend}.
+   * That is, if for any return statement there is some path such that the returned expression can be subtype of the expected type, then we
+   * do not raise an issue.
+   * Conversely, we raise an issue only if there is no way a returned expression could be (a subtype of) the expected type.
+   * (But we do not check the feasibility of paths.)
+   *
+   * Let us take a look at some relevant examples for illustration:
+   *
+   * <pre>
+   * class BoolMethodCheck:
+   *   def __init__(self, condition):
+   *     self.condition = condition
+   *
+   *   def __bool__(self):
+   *     return self.condition
+   * </pre>
+   *
+   * Here, {@code condition} can be anything and therefore technically it could not be a boolean.
+   * However, the user may only ever create instances of {@code BoolMethodCheck} where {@code self.condition} is a boolean.
+   * Hence, we do not report a rule violation in such cases as it might lead to too many FPs.
+   *
+   * The following case involves a Union type:
+   *
+   * <pre>
+   * def string_unknown() -> str:
+   *   ...
+   * class BoolMethodCheck:
+   *   def __bool__(self):
+   *     x = string_unknown()
+   *     if x == 'True':
+   *       x = True
+   *     elif x == 'False':
+   *       x = False
+   *   return x
+   * </pre>
+   *
+   * Here, the type of {@code x} is identified to be {@code Union[bool, str]} (simplified).
+   * In theory, this code may return a string but likely not in practice.
+   * Hence, we can not raise an issue just because a returned union type contains a type which is not a subtype of the expected type.
+   * (As long as the union type still contains another type that would be a valid return type.)
+   */
```

Comment:
I think this comment is not super relevant here.
You're describing the fact that sonar-python type inference is not path-sensitive and also that we prefer to under-approximate while raising an issue (i.e. we don't want to raise when the inferred type is "Any").

Those concepts are valid for all rules, so that's why I tend to think having a comment in the rule is not appropriate.

At the same time, I realize that from the eyes of a new maintainer, this comment would be useful.
So, I would suggest generalizing it a bit more and moving it to the `InferredType` interface.